packages/store/src/external/hydrate.js

```diff
@@ -0,0 +1,113 @@
+// @ts-check
+
+import '../types';
+
+import { makeWeakStore } from '../weak-store';
+import { makeStore } from '../store';
+
+/**
+ * @callback MakeBackingStore
+ * @param {HydrateHook} hydrateHook
+ * @returns {BackingStore}
+ */
+
+/**
+ * This creates an external store maker for a given storage backend, supporting
+ * the Closure interface that the rewriter targets.
+ *
+ * @template {Array<any>} A
+ * @template {ExternalInstance} T
+ * @param {MakeBackingStore} makeBackingStore
+ * @returns {MakeHydrateExternalStore<A, T>}
+ */
+export const makeHydrateExternalStoreMaker = makeBackingStore => {
+  /** @type {WeakStore<T, HydrateKey>} */
+  const instanceToKey = makeWeakStore('instance');
+
+  let lastStoreId = 0;
+
+  // This has to be a strong store, since it is indexed by ID.
+  const storeIdToHydrate = makeStore('storeId');
+
+  /**
+   * Create a data object that queues writes to the store.
+   *
+   * @param {HydrateData} data
+   * @param {() => void} markDirty
+   */
+  const makeActiveData = (data, markDirty) => {
+    const activeData = {};
+    // For every property in data...
+    for (const prop of Object.getOwnPropertyNames(data)) {
+      // Define a getter and setter on activeData.
+      Object.defineProperty(activeData, prop, {
+        get: () => data[prop],
+        set: value => {
+          data[prop] = value;
+          markDirty();
+        },
+      });
+    }
+    return harden(activeData);
+  };
+
+  /** @type {BackingStore} */
+  let backing;
+
+  /** @type {HydrateHook} */
+  const hydrateHook = {
+    getKey(value) {
+      return instanceToKey.get(value);
+    },
+    load([storeId, instanceId]) {
```

Comment:
Am I correct in thinking that liveslots will somehow get access to this `load` method, and will convert inbound references to our "virtual objects" into calls to `load()`?

Reply:
Correct.  Please see the next PR in this series, #1857.